Add "Signer Integration" page (#982)
casper-network/docs: source/docs/casper/developers/dapps/signer-integration.md

# Using the Casper Signer in dApp Development

The [Casper Signer](https://github.com/casper-ecosystem/signer) is a browser plugin for interfacing with Casper accounts. You can interact with the Casper Signer using the `Signer` class in the [Casper JavaScript SDK](https://github.com/casper-ecosystem/casper-js-sdk/).

## Integrating with the Casper Signer

To connect to the Casper Signer from a dApp's front-end is simple. Start by importing the `Signer` class from the Casper JS SDK.

Using ES5 syntax:

```javascript
const { Signer } = require("casper-js-sdk");
```

Using ES6:

```javascript
import { Signer } from "casper-js-sdk";
```

You can then use the `Signer` class to send a connection request to the Casper Signer:

```javascript
Signer.sendConnectionRequest()
```

> and check if the Casper Signer is connected using:

Reviewer: Start the sentence with a capital letter please:
```suggestion
Check if the Casper Signer is connected using the following:
```

Using ES5 syntax:

```javascript
const { Signer } = require("casper-js-sdk");
```

Using ES6:

```javascript
import { Signer } from "casper-js-sdk";
```

You can then use the `Signer` class to send a connection request to the Casper Signer:

```javascript
Signer.sendConnectionRequest()
```

and check if the Casper Signer is connected using:

```javascript
Signer.isConnected()
```

## Create Vaults & Accounts

You can use available JavaScript methods to create new vaults and accounts within the Signer. A vault is a wallet that may hold many accounts, whereas an account is an entity with a single public and private key that can interact with the Casper Network.

To check if the user has created a vault, read the boolean result of the promissory function :

```javascript
Signer.hasCreatedVault()
```

To create a new vault, call the following function, providing a password for the vault:

```javascript
Signer.createNewVault("password")
```

You may also create a test account that will remain persistent in the Signer using the function:

```javascript
Signer.createTestAccount("Account Name", privateKey)
```

> where `privateKey` is the base 16 private key of the account as a `string`

Reviewer 2: I think we should leave this. Without clarification, it is ambiguous as to whether one should use a base 16 key, base 64, bytes, string etc.

## Get Account Public Key

There are two available methods for getting the active account's public key, in different formats. There are many reasons you may want to read the public key of the active account, the main being the use it in a smart contract call or query. 

To get the account public key in hexadecimal format, execute:

```javascript
Signer.getActivePublicKey()
```

> This function resolves to a `string`, or rejects with an error.

Reviewer 2: In JavaScript, the `string` type is uncapitalized. Also, since `getActivePublicKey` is a promissory function, it does in fact `reject` not `return`.

> This function resolves to a `string`, or rejects with an error.

Reviewer: Sounds good, thanks!

To get the public key in base 64, run:

```javascript
Signer.getSelectedPublicKeyBase64()
```

Note that this function is deprecated in favor of `getActivePublicKey()`. It is recommended to use `getActivePublicKey()` and convert to base 64.

## Signing Messages and Transactions

### Sign Message

> When a message is signed with a Casper account, it serves as a proof that the message was sent by the owner of that account, and that the message has not been tampered with since it was signed. The signature is generated using the account's private key, which only the owner of the account should have access to. The signature can then be verified by anyone who has access to the message, the signature, and the public key associated with the Casper account.

Reviewer: ```suggestion
When a message is signed with a Casper account, it serves as proof that the account owner sent the message and that the message has not been tampered with since it was signed. The signature is generated using the account's private key, which only the account owner can access. The signature can then be verified by anyone with access to the message, the signature, and the public key associated with the Casper account.
```

> To sign a given message using the Casper JS SDK, initiate a signature request using this function:

Reviewer: ```suggestion
To sign a given message using the Casper JS SDK, initiate a signature request using this function, where `publicKey` is the Base16 `String`-typed public key of the active account within the Casper Signer.
```

```javascript
Signer.signMessage("Message to be signed", publicKey)
```

> where `publicKey` is the base 16 `string`-typed public key of the active account within the Casper Signer.

Reviewer: See above
```suggestion

```

### Sign Transaction

> Transactions can be forwarded to the Casper Signer to be signed by a user without the user needing to reveal their private key to the application. To accomplish this, convert a valid, unsigned `Deploy` object to JSON and call the `sign` method on the `Signer` class:

Reviewer: ```suggestion
Transactions can be forwarded to the Casper Signer to be signed by a user without revealing their private key to the application. To accomplish this, convert a valid, unsigned `Deploy` object to JSON and call the `sign` method on the `Signer` class:
```

```javascript
const publicKeyHex = await Signer.getActivePublicKey()

const unsignedDeploy = contract.callEntrypoint(
  "test_entrypoint",
  runtimeArguments,
  CLPublicKey.fromHex(publicKeyHex),

>   "casper", // Or "casper-test" for testnet

Reviewer: ```suggestion
  "casper", // Or "casper-test" for Testnet
```

The Casper Signer will emit events when it changes states. By listening for these events, you can update your application to respond to them.

For example, the Casper Signer will lock itself automatically after a set interval of disuse (2-10 minutes). When this occurs, the Casper Signer emits the `signer:locked` event. We can listen for this event by using `addEventListener` on the `window`.

```javascript
window.addEventListener("signer:locked", (msg) => {
  // Disable buttons, forget the active public key, etc
});
```

The variable `msg` shown above contains event information specific to the event such as `isConnected` `isUnlocked` and the active public key.

There are a few different events you may listen to:

* `signer:connected`

  Emitted when the Casper Signer connects to the dApp.

* `signer:disconnected`

  Emitted when the Casper Signer disconnects from the dApp.

* `signer:tabUpdated`

  Emitted when a tab is updated within the Signer.

* `signer:activeKeyChanged`

  Emitted when the active account changes within the Signer.

* `signer:locked`

  Emitted when the Casper Signer is locked. Can be performed by the user or occur due to timeout.

* `signer:unlocked`

>   Emitted when the Casper Signer is unlocked.

Reviewer: Please reformat this list to match other lists in the docs. For example, one bullet could be:

* `signer:unlocked` -  Emitted when the Casper Signer is unlocked

## Disconnecting

> After connecting the Casper Signer to your dApp, you may at some point want to disconnect. There is a method available for this:

Reviewer: ```suggestion
After connecting the Casper Signer to your dApp, you may, at some point, want to disconnect. There is a method available for this:
```
